### PR TITLE
feat(command): add support to delete dependants of command resource

### PR DESCRIPTION
**Pull request by mittachaitu**
This PR does the following changes:
- Add support to delete dependent resources when
  command resource is marked for deletion.
- Add support to launch the jobs periodically when
  the command is configured for run always

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

### CODE REVIEW

File: pkg/command/reconciler.go [hunk 1]
```diff
@@ -180,21 +195,45 @@ func (r *Reconciliation) initChildJobDetails() {
...
+				r.childJob.GetKind(),
+				r.childJob.GetNamespace(),
+				r.childJob.GetName(),
+			)
+			return
+		}
+		condType := condition["type"].(string)
+		if condType == types.JobPhaseCompleted {
+			condStatus := condition["status"].(string)
+			if condStatus == "True" {
```
Comment (AmitKumarDas) [inline pkg/command/reconciler.go:227]: to be more safe lets compare by doing a toLowerCase() == "true"

File: pkg/command/reconciler.go [hunk 2]
```diff
@@ -180,21 +195,45 @@ func (r *Reconciliation) initChildJobDetails() {
...
+				r.isChildJobCompleted = true
```
Comment (AmitKumarDas) [inline pkg/command/reconciler.go:228]: As per this logic if any condition status is True we mark the job as completed.
Is that okay?

File: pkg/command/reconciler.go [hunk 3]
```diff
@@ -445,12 +488,28 @@ func (r *Reconciliation) reconcileRunAlwaysCommand() (types.CommandStatus, error
...
+		_, err := r.deleteChildJob()
+		if err != nil {
+			return types.CommandStatus{}, err
+		}
+		klog.V(1).Infof("Deleted command job: Command %q / %q successfully",
+			r.command.GetNamespace(),
+			r.command.GetName(),
+		)
+
+		klog.V(1).Infof(
```
Comment (AmitKumarDas) [inline pkg/command/reconciler.go:507]: Can we mention as code comments etc that we dont want this to be eventual but immediate. We can explain that delete & create will happen in the same reconciliation process.

Comment (AmitKumarDas) [reply pkg/command/reconciler.go:507]: Have we double checked if there will be no conflict during delete & immediate create.  The job name might conflict?
We might want to generate the job name or we might want to retry job create in some loop.

File: pkg/command/reconciler.go [hunk 4]
```diff
@@ -115,7 +124,12 @@ func (r *Reconciliation) initChildJobDetails() {
 		)
 		return
 	}
-	got, found, err := r.isChildJobAvailable()
+
+	if r.isChildJobAvailableFn != nil {
+		got, found, err = r.isChildJobAvailableFn()
+	} else {
+		got, found, err = r.isChildJobAvailable()
```
Comment (AmitKumarDas) [inline pkg/command/reconciler.go:131]: I am rethinking the naming part. IMO getChildJob() & getChildJobFn() will be better names.
The rest of signature can remain same.

File: pkg/command/reconciler_test.go [hunk 5]
```diff
@@ -30,98 +32,154 @@ import (
...
-// 				t.Fatalf(
-// 					"Expected child job as completed %t got %t",
-// 					mock.ExpectCompletedChildJob,
-// 					r.isChildJobCompleted,
-// 				)
-// 			}
-// 		})
-// 	}
-// }
+func fakeChildAvailable(
```
Comment (AmitKumarDas) [inline pkg/command/reconciler_test.go:35]: We can rename this to getFakeChildJob(....)
Please have a look at my previous comments on renaming.

Comment (AmitKumarDas) [reply pkg/command/reconciler_test.go:35]: I really liked the way you have understood the logic & intention so quickly!!!
:bow: @mittachaitu :+1: :100:

File: config/metac.yaml [hunk 6]
```diff
@@ -64,4 +64,51 @@ spec:
...
+  - apiVersion: v1
+    resource: configmaps
+    advancedSelector:
+      selectorTerms:
+      # select ConfigMap if its labels has following
+      - matchLabels:
+          command.dope.mayadata.io/lock: "true"
+        matchReferenceExpressions:
+        - key: metadata.labels.command\.dope\.mayadata\.io/name
+          operator: EqualsWatchName # match this lbl value against watch Name
```
Comment (AmitKumarDas) [inline config/metac.yaml:109]: These are my thoughts @mittachaitu 
- Job may have auto generated name
- Job will have Command's uid in its label
- Job will have Command's name in its label
- ConfigMap may have either auto generated name or named as cmd-<command's uid>
- ConfigMap will have Command's uid in its label
- ConfigMap will have Command's name in its label

The finalizer selector can make use of name & uid matches accordingly. I guess relying on name is not enough. It should be name & namespace. However, we might want to stick to UID for selection. Label key should be namespaced e.g. "command.dope.mayadata.io/name", "command.dope.mayadata.io/uid", etc.

Use of name selector might seem to be unnecessary since uid is sufficient. However, name will help to exclude the resource from the list quickly & hence improve the select performance. (_I hope the metac advanced selector matches by iterating the list of selectors. If name is the first selector then rejection will be fast._)

Comment (AmitKumarDas) [reply config/metac.yaml:109]: @mittachaitu try to think deep on this!!!

File: types/command/command.go [hunk 7]
```diff
@@ -89,7 +89,7 @@ const (
 const (
 	// JobPhaseCompleted represents a Kubernetes Job with
 	// Completed status
-	JobPhaseCompleted string = "Completed"
+	JobPhaseCompleted string = "Complete"
```
Comment (AmitKumarDas) [inline types/command/command.go:92]: I think there are edge cases. We should try to write the integration tests & see how things work.
We can team up with @amitbhatt818 or others if you want.

Comment (mittachaitu) [reply types/command/command.go:92]: I am on it writing integration test cases

File: pkg/command/reconciler.go (hunk 2 above)
Comment (mittachaitu) [reply pkg/command/reconciler.go:228]: No, we are marking the job as completed only when the condition type is [Complete](https://github.com/kubernetes/kubernetes/blob/b1a2a99f1fe1622d20b68abaa99348e19a35c704/pkg/apis/batch/types.go#L194) and status is set to `True`

File: pkg/command/reconciler.go (hunk 3 above)
Comment (mittachaitu) [reply pkg/command/reconciler.go:507]: Hmm, good catch!! Instead, I will wait till the object successfully deleted from etcd

File: types/command/command.go (hunk 7 above)
Comment (mittachaitu) [reply types/command/command.go:92]: As per code flow, I am not seeing any edge case except deletion & creation in the same reconciliation

File: config/metac.yaml [hunk 8]
```diff
@@ -64,4 +64,54 @@ spec:
...
+    resource: commands
+  attachments:
+  # Delete pod
+  - apiVersion: v1
+    resource: pods
+    advancedSelector:
+      selectorTerms:
+      # select pod if its namespace is same as watch namespace
+      - matchFields:
+          metadata.namespace: EqualsWatchNamespace
```
Comment (AmitKumarDas) [inline config/metac.yaml:86]: Can you please check if EqualsWatchNamespace is really supported for matchFields?
Please check what's the supported key & value as well.
https://github.com/AmitKumarDas/metac/blob/master/apis/metacontroller/v1alpha1/selector.go

Comment (AmitKumarDas) [reply config/metac.yaml:86]: https://github.com/AmitKumarDas/metac/tree/master/controller/common/selector
https://github.com/AmitKumarDas/metac/blob/master/controller/common/selector/reference_test.go

File: controller/command/finalizer.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+// NOTE:
+// 	SyncHookRequest is the payload received as part of finalize
+// request. Similarly, SyncHookResponse is the payload sent as a
+// response as part of finalize request.
+//
+// NOTE:
+//	Returning error will panic this process. We would rather want this
+// controller to run continuously. Hence, the errors are handled.
+func Finalize(request *generic.SyncHookRequest, response *generic.SyncHookResponse) error {
+	klog.Infof("[Debug] In Finalize hook attach lengths %d", request.Attachments.Len())
```
Comment (AmitKumarDas) [inline controller/command/finalizer.go:48]: [Debug] In finalize hook: Attachments %d

File: controller/command/finalizer.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	if request.Attachments.IsEmpty() {
+		// Since no Dependents found it is safe to delete Command
+		response.Finalized = true
+		return nil
+	}
+
+	response.ResyncAfterSeconds = defaultDeletionResyncTime
+	response.ExplicitDeletes = request.Attachments.List()
```
Comment (AmitKumarDas) [inline controller/command/finalizer.go:56]: Can we add a comment that all the observed attachments will get deleted

File: controller/command/reconciler.go [hunk 11]
```diff
@@ -61,6 +61,17 @@ func (r *Reconciler) eval() {
...
 func (r *Reconciler) sync() {
+	// Check for deletion timestamp on command resource
+	// if set then command is marked for deletion
+	if !r.observedCommand.DeletionTimestamp.IsZero() {
+		klog.V(1).Infof(
+			"Will skip command reconciliation: It is marked for deletion: Command %q / %q",
+			r.observedCommand.GetNamespace(),
+			r.observedCommand.GetName(),
+		)
+		return
```
Comment (AmitKumarDas) [inline controller/command/reconciler.go:72]: There is a specific variable that can be set to skip reconcile.
Lets set that.
Please check the embedded reconciler properties & logic as well.

File: controller/command/finalizer.go (hunk 9 above)
Comment (mittachaitu) [reply controller/command/finalizer.go:48]: This will be removed... I just added for debugging purpose

File: test/declarative/experiments/command-creation-deletion.yaml [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+                      - -v=3
+                        --command-name=testing-command
+                        --command-ns=recipe-integration-cmd-testing
+  - name: assert-cm-lock-creation
+    assert:
+      state:
+        kind: ConfigMap
+        apiVersion: v1
+        metadata:
+          name: testing-command-lock
```
Comment (AmitKumarDas) [inline test/declarative/experiments/command-creation-deletion.yaml:105]: are we going to change the lock name in this PR?
e.g. Will it make use of uid of Command or use generateName property

File: config/metac.yaml (hunk 8 above)
Comment (mittachaitu) [reply config/metac.yaml:86]: Not aware!! I looked at API and written this. There is no information written in APIs

Comment (mittachaitu) [reply config/metac.yaml:86]: https://github.com/AmitKumarDas/metac/blob/master/apis/metacontroller/v1alpha1/selector.go -- API link

File: controller/command/reconciler.go (hunk 11 above)
Comment (mittachaitu) [reply controller/command/reconciler.go:72]: I don't think it will stop reconciliation on the command but it says [skip](https://github.com/AmitKumarDas/metac/blob/master/controller/generic/hooks.go#L94-L96) reconciliation on attachments

Comment (mittachaitu) [reply controller/command/reconciler.go:72]: We are already setting it https://github.com/mayadata-io/d-operators/blob/master/controller/command/reconciler.go#L123-L125

File: test/declarative/experiments/command-creation-deletion.yaml (hunk 12 above)
Comment (mittachaitu) [reply test/declarative/experiments/command-creation-deletion.yaml]: Updated

File: controller/command/finalizer.go (hunk 10 above)
Comment (mittachaitu) [reply controller/command/finalizer.go:57]: Added the review comment

File: pkg/command/reconciler.go (hunk 3 above)
Comment (mittachaitu) [reply pkg/command/reconciler.go:532]: Updated PTAL

File: config/metac.yaml (hunk 6 above)
Comment (mittachaitu) [reply config/metac.yaml]: make use of uid to test changes

File: config/metac.yaml (hunk 8 above)
Comment (mittachaitu) [reply config/metac.yaml]: Updated

File: controller/command/reconciler.go (hunk 11 above)
Comment (AmitKumarDas) [reply controller/command/reconciler.go:72]: ok. perfect.

**Review by AmitKumarDas — APPROVED** [review]
